docs/proposals/multi-tenant-networking.md

```diff
@@ -0,0 +1,194 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/multi-tenant-networking.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Multi-tenant networking
+
+## Abstract
+
+A proposal of multi-tenant networking by network providers.
+
+## Motivation
+
+The current kubernetes network design works great for connecting containers and building services, but it doesn’t work on many use-cases, such as container network isolation, ip address overlapping and so on. So this proposal adds network provider which adds multi-tenant networking for kubernetes.
+
+Goals of this design:
+
+* Add multi-tenant networking to support network isolation and ip address overlapping between containers
```

Comment:
This would limit you from being able to join any arbitrary network with one another, whether you are to combine two or more networks into one larger network (due to organisational changes) or introduce a new network that is to be attached to a running application. The question is, of course, whether this may be considered a kind limitation that may not be introduced or it's something nobody cares about and infrastructures should be immutable etc?